File: haskell/src/function.hs
{-# LANGUAGE OverloadedStrings, DeriveGeneric #-}

import qualified Data.ByteString.Lazy as B
import Data.Aeson
import Data.List
import GHC.Generics

main = undefined

get :: IO [Transaction]
get = do
  d <- (eitherDecode <$> getJSON) :: IO (Either String [Transaction])
  case d of
    Left err -> error "Erro ao decodificar"
    Right ts -> return ts

data Date = Date { year :: Int
                 , month :: Int
                 , dayOfMonth :: Int
                 } deriving (Show, Generic)

data Transaction = Transaction { date :: Date
                               , textoIdentificador :: String
                               , valor :: Float
                               , descricao :: String
                               , numeroDOC :: String
                               , classificada :: Bool
                               , tipos :: [String]
                               , arquivos :: [String]
                               } deriving (Show, Generic)


instance FromJSON Date
instance ToJSON Date
instance FromJSON Transaction
instance ToJSON Transaction


jsonFile :: FilePath
jsonFile = "../../data/dataset.json"

getJSON :: IO B.ByteString
getJSON = B.readFile jsonFile


-- Filter transactions by year
filterByYear :: [Transaction] -> Int -> [Transaction]
filterByYear ts y = filter (\t -> checkYear t y) ts

-- Filter transactions by month and year
filterByYearMonth :: [Transaction] -> Int -> Int -> [Transaction]
filterByYearMonth ts y m = filter (\t -> checkYear t y && checkMonth t m) ts

-- Calculate income by month and year
income :: [Transaction] -> Int -> Int -> Float
income ts y m = foldl (+) 0 $ map valor $ filterIncome
                where filterIncome = filter (\t -> validTransaction t && checkYear t y && checkMonth t m && valor t > 0) ts

-- Calculate expenses by month and year
expenses :: [Transaction] -> Int -> Int -> Float
expenses ts y m = foldl (+) 0 $ map valor $ filterExpenses
                  where filterExpenses = filter (\t -> validTransaction t && checkYear t y && checkMonth t m && valor t < 0) ts

-- Calculate net income by month and year
netIncome :: [Transaction] -> Int -> Int -> Float
netIncome ts y m = foldl (+) 0 $ map valor $ filterByYearMonth ts y m

-- Calculate balance by month and year
balance :: [Transaction] -> Int -> Int -> Float
balance ts y m = initialBalance ts y m + netIncome ts y m

-- Calculate maximum balance in a specified month and year
maximumBalance :: [Transaction] -> Int -> Int -> Float
maximumBalance ts y m = maximum $ map (snd) $ cashFlow ts y m

-- Calculate minimum balance in a specified month and year
minimumBalance :: [Transaction] -> Int -> Int -> Float
minimumBalance ts y m = minimum $ map (snd) $ cashFlow ts y m

-- Calculate average income by year
averageIncome :: [Transaction] -> Int -> Float
averageIncome ts y = (foldl (+) 0 $ [valor t | t <- (filterByYear ts y), valor t > 0]) / 12

-- Calculate average expenses by year
averageExpenses :: [Transaction] -> Int -> Float
averageExpenses ts y = (foldl (+) 0 $ [valor t | t <- (filterByYear ts y), valor t < 0]) / 12

-- Calculate average net income by year
averageNetIncome :: [Transaction] -> Int -> Float
averageNetIncome ts y = (foldl (+) 0 $ map valor $ filterByYear ts y) / (12)

-- Return cash flow in a specified month/year
cashFlow :: [Transaction] -> Int -> Int -> [(Int, Float)]
cashFlow ts y m = [(d, (initialBalance ts y m) + dailyBalance ts y m d) | d <- [1..(1 + (dayOfMonthsInMonth m))]] 

-- Auxiliary functions
--
dailyBalance :: [Transaction] -> Int -> Int -> Int -> Float
dailyBalance ts y m d = sum $ map (valor) $ filter (\t -> checkYear t y && checkMonth t m && (dayOfMonth . date) t <= d && validTransaction t) ts 

checkYear :: Transaction -> Int -> Bool
checkYear t y = (year . date) t == y

checkMonth :: Transaction -> Int -> Bool
checkMonth t m = (month . date) t == m

checkdayOfMonth :: Transaction -> Int -> Bool
checkdayOfMonth t d = (dayOfMonth . date) t == d

validTransaction :: Transaction -> Bool
validTransaction t = let saldoCorrenteCheck = "SALDO_CORRENTE" `elem` (tipos t)
                         aplicacaoCheck = "APLICACAO" `elem` (tipos t)
                         valorAplicacaoCheck = "VALOR_APLICACAO" `elem` (tipos t)
                     in not saldoCorrenteCheck && not aplicacaoCheck && not valorAplicacaoCheck

initialBalance :: [Transaction] -> Int -> Int -> Float
initialBalance ts y m = let result = Data.List.find (\t -> checkYear t y && checkMonth t m && (dayOfMonth . date) t == 1 && "SALDO_CORRENTE" `elem` (tipos t)) ts
                        in case result of
                          Just x -> valor x
                          Nothing -> error "This wasn't supposed to happen"

dayOfMonthsInMonth :: Int -> Int
dayOfMonthsInMonth m = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31] !! m

File: javascript/server/src/function.js
'use strict';

/*
 * Implementation in functional Javascript
 *
 * @author Gustavo Alves
 */

function filterByYear (transactions, year) {
  return transactions.filter(t => t.data.year == year);
};

function filterByMonth (transactions, year, month) {
  return transactions.filter(t => t.data.year == year && t.data.month == month);
};

function income (transactions, year, month) {
  const filtered = transactions.filter(t => {
    return t.data.year == year &&
      t.data.month == month &&
      t.valor > 0 &&
      !t.tipos.includes('SALDO_CORRENTE') &&
      !t.tipos.includes('APLICACAO') &&
      !t.tipos.includes('VALOR_APLICACAO');
  });
  return filtered.reduce(((acc, t) => acc + t.valor), 0);
};

function expenses (transactions, year, month) {
  const filtered = transactions.filter(t => {
    return t.data.year == year &&
      t.data.month == month &&
      t.valor < 0 &&
      !t.tipos.includes('SALDO_CORRENTE') &&
      !t.tipos.includes('APLICACAO') &&
      !t.tipos.includes('VALOR_APLICACAO');
  });
  return filtered.reduce(((acc, t) => acc + t.valor), 0);
};

function netIncome (transactions, year, month) {
  const filtered = transactions.filter(t => {
    return t.data.year == year &&
      t.data.month == month &&
      !t.tipos.includes('SALDO_CORRENTE') &&
      !t.tipos.includes('APLICACAO') &&
      !t.tipos.includes('VALOR_APLICACAO');
  });
  return filtered.reduce(((acc, t) => acc + t.valor), 0);
};

function balance (transactions, year, month) {
  const initialBalance = startingBalance(transactions, year, month);
  const nettoMonth = netIncome(transactions, year, month);
  return initialBalance + nettoMonth;
};

function averageIncome (transactions, year) {
  const filtered = transactions.filter(t => {
    return t.data.year == year &&
      t.valor > 0 &&
      !t.tipos.includes('SALDO_CORRENTE') &&
      !t.tipos.includes('APLICACAO') &&
      !t.tipos.includes('VALOR_APLICACAO');
  });
  const nom = numberOfMonths(transactions, year);
  return filtered.reduce(((acc, t) => acc + t.valor), 0) / nom;
};

function averageExpense (transactions, year) {
  const result = transactions.filter(t => {
    return t.data.year == year &&
      t.valor < 0 &&
      !t.tipos.includes('SALDO_CORRENTE') &&
      !t.tipos.includes('APLICACAO') &&
      !t.tipos.includes('VALOR_APLICACAO');
  });
  const nom = numberOfMonths(transactions, year);
  return result.reduce(((acc, t) => acc + t.valor), 0) / nom;
};

function averageNetIncome (transactions, year) {
  const result = transactions.filter(t => {
    return t.data.year == year &&
      !t.tipos.includes('SALDO_CORRENTE') &&
      !t.tipos.includes('APLICACAO') &&
      !t.tipos.includes('VALOR_APLICACAO');
  });
  const nom = numberOfMonths(transactions, year);
  return result.reduce(((acc, t) => acc + t.valor), 0) / nom;
};

function cashFlow (transactions, year, month) {
  const filtered = filterByMonth(transactions, year, month);
  const ldm = lastDayOfMonth(year, month);
  const cf = [];
  cf[0] = {
    day: 1,
    balance: startingBalance(transactions, year, month) + dailyNetto(filtered, year, month, 1)
  };
  for (let i = 1; i < ldm + 1; i++) {
    cf[i] = {
      day: i + 1,
      balance: cf[i - 1].balance + dailyNetto(filtered, year, month, i + 1)
    };
  };
  return cf;
};

function maximumBalance (transactions, year, month) {
  const dailyBalances = cashFlow(transactions, year, month);
  const reducer = (acc, b) => (b.balance > acc.balance)? b : acc;
  return dailyBalances.reduce(reducer, dailyBalances[0]);
};

function minimumBalance (transactions, year, month) {
  const dailyBalances = cashFlow(transactions, year, month);
  const reducer = (acc, b) => (b.balance < acc.balance)? b : acc;
  return dailyBalances.reduce(reducer, dailyBalances[0]);
}

function startingBalance (transactions, year, month) {
  const result = transactions.find(t => {
    return t.data.year == year &&
      t.data.month == month &&
      t.data.dayOfMonth == 1 &&
      t.tipos.includes('SALDO_CORRENTE');
  });
  return result.valor;
};

function numberOfMonths (transactions, year) {
  let min = 11;
  let max = 0;
  
  transactions.forEach(t => {
    if (t.data.month < min) min = t.data.month;
    if (t.data.month > max) max = t.data.month; 
  });
  return max + min + 1;
};

function lastDayOfMonth(year, month) {
  let lastDay;
  const months = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  
  if (month == '01' && (year % 100 != 0 && year % 4 == 0 || year % 400 == 0)) lastDay = 29;
  else lastDay = months[parseInt(month)];
  return lastDay;
}

function dailyNetto (transactions, year, month, day) {
  const filtered = transactions.filter(t => {
    return t.data.year == year &&
      t.data.month == month &&
      t.data.dayOfMonth == day &&
      !t.tipos.includes('SALDO_CORRENTE') &&
      !t.tipos.includes('APLICACAO') &&
      !t.tipos.includes('VALOR_APLICACAO');
  });
  return filtered.reduce(((acc, t) => acc + t.valor), 0);
}

module.exports = {
  filterByYear,
  filterByMonth,
  income,
  expenses,
  netIncome,
  balance,
  maximumBalance,
  minimumBalance,
  averageIncome,
  averageExpense,
  averageNetIncome,
  cashFlow
}